libs/pilight/protocols/433.92/rev_v5.c

```diff
@@ -0,0 +1,414 @@
+		if(binary[i]==1) {
+			x=i*4;
+			createHigh(16+x, 16+x+3); //first unit pulse is 16
+		}
+	}
+}
+
+
+static void createState(bool on, bool all) {
+	//based on rev_v3.c
+	int binary[255];
+	int length = 0;
+	int i=0, x=0;
+    int rawState;
+
+	if(all) {
+		rawState = on ? 0 : 3; //ALL-ON -> 0 //ALL-OFF -> 3
+	}
+	else {
+		rawState = on ? 1 : 2; //ON -> 1 //OFF -> 2
+	}
+	
+	length = decToBinRev(rawState, binary);
+
+	//logprintf(LOG_INFO, "rev_v5:sending state value '%d', bit-length:%d", rawState, length);
+
+	for(i=0;i<=length;i++) {
+		if(binary[i]==1) {
+			x=i*4;
+			createHigh(32+x, 32+x+3); //first unit pulse is 32
+		}
+	}
+}
+
+static void createPostfix(void) {
+	//the last 2 bits are always set
+	createHigh(40, 40+3); 
+	createHigh(44, 44+3); 
+}
+
+static void createFooter(void) {
+	//copied from rev_v3.c
+	rev5_switch->raw[48]=(AVG_PULSE_LENGTH);
+	rev5_switch->raw[49]=(PULSE_DIV*AVG_PULSE_LENGTH);
+}
+
+static int createCode(struct JsonNode *code) {
+	char sId[4] = {'\0'};
+	char *stmp = NULL;
+	double itmp = -1;
+	int id = -1;
+	int unit = -1;
+	int state = -1;
+	int all = 0;
+	int areabit = 0;
+
+	strcpy(sId, "-1");
+
+	if(json_find_string(code, "id", &stmp) == 0)
+		strcpy(sId, stmp);
+
+	if((int)(sId[0]) >= 48 /*0*/ && (int)(sId[0]) <= 57 /*9*/) {
+		//Id is already a number
+		id = atoi(sId);
+	}
+    else if((int)(sId[0]) >= 65 /*A*/ && (int)(sId[0]) <= 80 /*P*/) {
+		//Id is defined by char
+		id = createIdFromLetter((int)sId[0]);
+	}
+	else {
+		logprintf(LOG_ERR, "rev5_switch: invalid id number specified");
+	}
+	
+
+	if(json_find_number(code, "unit", &itmp) == 0)
+		unit = (int)round(itmp);
+	if(json_find_number(code, "all", &itmp) == 0)
+		all = (int)round(itmp);
+	if(json_find_number(code, "off", &itmp) == 0)
+		state=0;
+	if(json_find_number(code, "on", &itmp) == 0)
+		state=1;
+
+	if((id == -1) || (state == -1) || (unit == -1 && all == 0)) { //id is not set, state is not set, neither unit nor all is set
+		logprintf(LOG_ERR, "rev5_switch: insufficient number of arguments");
+		return EXIT_FAILURE;
+	} else if(id > (ID_COUNT-1) || id < 0) {
+		logprintf(LOG_ERR, "rev5_switch: invalid id range");
+		return EXIT_FAILURE;
+	} else if(unit > -1 && all > 0) {
+		logprintf(LOG_ERR, "rev5_switch: only unit-parameter or all-parameter can be specified");
+		return EXIT_FAILURE;
+	} else if((unit > 7 || unit < 0) && (all == 0)) {
+		logprintf(LOG_ERR, "rev5_switch: invalid unit range");
+		return EXIT_FAILURE;
+	} else if((unit == -1) && (all > 2 || (all < 1))) {
+		logprintf(LOG_ERR, "rev5_switch: invalid all value (use 1 for units 0-3 and 2 for units 4-7)");
+		return EXIT_FAILURE;
+	}
+	else {
+		if (unit >= 0)
+			areabit = (unit > 3) ? 1 : 0; 
+			
+		if(all > 0) {
+			unit = 0; //broadcast uses first device per group
+			areabit = all-1;
+		}
+
+		createMessage(id, unit, state, all);
+		clearCode();
+		createId(id);
+		createUnit(generateUnitValue(unit, areabit));
+		createState(state > 0, all > 0);
+		createPostfix();
```

Comment:
Why don't you move this code into the `clearCode` function?

Reply:
This flow is already used by the rev_v3.c protocol. I tried to keep the same structure as existing rev protocols because they might be combined one day.
`clearCode` just resets all pulses to zeros and these methods fill the pulses with ones again. I whould need to pass `id/unit/areabit/state/all` to 'clearCode` just for calling the other methods. I do not see the benefit you see.

Reply:
Ok